### DISCUSSION TITLE
Why use `/_doc/_mget` as request path when try to multi get documentation?

### DISCUSSION THREAD

**Discussion by ivyxjc** [apache/skywalking#9796] [Q&A] [answered]
[ElasticSearch api documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-multi-get.html) informs that multi-get request path is `/_mget`. But in `V7DocumentFactory`, it use the path `/_doc/_mget`. 

https://github.com/apache/skywalking/blob/04c4fedf436c35b326e7604e4a86af9134dab13c/oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/requests/factory/v7plus/V7DocumentFactory.java#L112-L115

And in AWS OpenSearch, it will return `{"Message":"Your request: '/_doc/_mget' is not allowed."}`.

**Comment by wu-sheng** [comment]
@kezhenxu94

**Comment by kezhenxu94** [comment]
@ivyxjc can you share which OpenSearch version are you using?

**Reply by ivyxjc** [reply]
AWS-host opensearch 1.3.
I just checked in self-host opensearh (run with docker), `/_doc/_mget` is ok. But in aws host opensearch, it will return the error response.

**Reply by wu-sheng** [reply]
I think this seems something AWS changed privately? 
OpenSearch actually is a fork of ElasticSearch 7.10 or another 7.x minor version. We could only tested with open source versions.

**Reply by wu-sheng** [reply]
@ivyxjc It is better you could ask OpenSearch Community, they should have a lot of AWS engineers.

**Comment by weiguo-tough** [comment]
We also encountered the same problem, our solution is to change to "/_mget", build a private image, and then replace the image in the helm install configuration.

**Comment by ivyxjc** [comment, marked as answer]
Both [elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-multi-get.html)/[opensearch](https://opensearch.org/docs/latest/api-reference/document-apis/multi-get/) multi get documentation only have the following two apis
```
GET _mget
GET <index>/_mget
```
For path /_doc/_mget , Opensearch/Elastisearch should see `_doc` as index name, but `_doc` cannot be index name since index name should not start with `_`.

Opensearch/Elasticsearch real action has some conflicts with documentation. I recommend we'd better following the documentation.


I raise one [related issue](https://github.com/opensearch-project/OpenSearch/issues/4903) in opensearch project

**Reply by wu-sheng** [reply]
Thanks for the update. We would love to see what is the final resolution about the APIs. And how we could know the difference from the client side. 

So, is there an OpenSearch version to reproduce this error?

**Reply by ivyxjc** [reply]
Actually, only `AWS hosted opensearch/elasticsearch` will see  path `_doc/_mget` as illegal path. Opensource opensearch or elasticsearch all accept `_doc/_mget` call. I think that `AWS hosted opensearch/elasticsearch` is the only one which strict compliance with the documentation

**Reply by wu-sheng** [reply]
Is there a version or something we could know which service is at peer?
Because I am not sure how to fix this for making all sides work.

**Reply by ivyxjc** [reply]
I think that changing the path from `/_doc/_mget` to `'/_mget` may can make all sides work. But it may need some more test.

**Reply by wu-sheng** [reply]
The key point is the following docs. Some compatible ways may be provided and may be removed without any warning.
If this specific path is required by AWS-hosted OpenSearch, we should only consider adopting it in that case only. Especially, you mentioned, the current way is correct for ElasticSearch and OpenSearch. A cloud-vendor hosted version is a black box.

**Reply by ivyxjc** [reply]
Current way can work correctly for ElasticSearch and OpenSearch, but it uses one way that is not listed in OpenSearch/ElasticSearch documentation. Recommend to use the same-function way listed in the documentation.

**Reply by wu-sheng** [reply]
>  but it uses one way that is not listed in OpenSearch/ElasticSearch documentation.

Could you be more clear about this? Such as what we are using, what is doc saying, and the doc ref? This could help us to adjust codes.

**Reply by ivyxjc** [reply]
Now skywalking are using path`_doc/_mget`:
https://github.com/apache/skywalking/blob/f9e23eb32911b85a90e2d7699ee4591dd11ad1a4/oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/requests/factory/v7plus/V7DocumentFactory.java#L112-L115

[docs-multi-get](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-multi-get.html) does not have documenation about path `_doc/_mget`

**Reply by wu-sheng** [reply]
@kezhenxu94 Could you recheck? I think consistent with docs is better. Be careful about version differences.

**Comment by kezhenxu94** [comment]
For record, this happens for AWS-hosted ElasticSearch 7.10 too, and should be fixed by https://github.com/apache/skywalking/pull/10031

**Reply by wu-sheng** [reply]
FYI @ivyxjc

**Reply by ivyxjc** [reply]
We deploy with the version(9.3.0-7ad0406), and it works perfect.
